### PR TITLE
Add user location feature in the map

### PR DESCRIPTION
**Pull request by aminebenji04**
This PR adds the user location functionality for the map, this includes new permissions, dependency injection updates and UI improvements.
### Summary of Changes

1.  Location Permissions and Checks
     - Implemented location permissions checking in MainActivity
     - Improved the naming of the other permissions to coexist with new permissions
2. User Location:
     - Updated LocationViewModel along with the repository to incorporate a function that calls the repository to fetch location data using the FusedLocation which is an API that permits to communicate with the phone and obtain location info, see more at https://developers.google.com/location-context/fused-location-provider?hl=fr
3. Permissions Handling Function
     - Implemented a very useful function "LocationPermissionChecker" that helps us to check the state of the permissions at any time, and this can be very gelpful for mocking tests, for example: mock LocationPermissionChecker to return false when you want to simulate a user that didn't accept to share his location. This logic can be used for all the permissions and for permissions that we already done like camera permissions.
4. UI enhancements
     - Added a button to re-center the camera on our current location with very responsive and fun animation.
5. Testing 
     - Added new unit tests and Android UI tests to cover location-related functions and ensure functionality across different scenarios
### How to try the new feature

1.  Once you open the app, you will be asked to accept to share your location, select fine location
2. Then you can go on the Map screen
3. You will see that your current position is marked with a blue marker
4. You can try to play aroung by going far from your position and clicking on the re-center button

### CODE REVIEW

File: app/src/test/java/com/android/sample/model/map/LocationViewModelTest.kt [hunk 1]
```diff
@@ -47,4 +50,69 @@ class LocationViewModelTest {
     // Then
     coVerify(exactly = 0) { mockRepository.search(any(), any(), any()) }
   }
+
```
Comment (DAFERMOH) [inline app/src/test/java/com/android/sample/model/map/LocationViewModelTest.kt:53]: The tests effectively cover both scenarios when permissions are granted and denied, which is essential for verifying location access logic. The use of coVerify(exactly = 0) to check that getCurrentLocation is not called when permission is denied is a good approach.

File: app/src/test/java/com/android/sample/model/map/NominatimLocationRepositoryTest.kt [hunk 2]
```diff
@@ -1,40 +1,52 @@
...
+  @Mock private lateinit var fusedLocationProviderClient: FusedLocationProviderClient
   @Mock private lateinit var mockCall: Call
+  @Mock private lateinit var mockLocationTask: Task<AndroidLocation>
 
   private lateinit var locationRepository: NominatimLocationRepository
 
   @Before
   fun setUp() {
     MockitoAnnotations.openMocks(this)
-    locationRepository = NominatimLocationRepository(mockClient)
```
Comment (DAFERMOH) [inline app/src/test/java/com/android/sample/model/map/NominatimLocationRepositoryTest.kt:32]: LGTM!

File: app/src/test/java/com/android/sample/model/map/NominatimLocationRepositoryTest.kt [hunk 3]
```diff
@@ -124,6 +136,35 @@ class NominatimLocationRepositoryTest {
     assertTrue(failureCalled)
   }
```
Comment (DAFERMOH) [inline app/src/test/java/com/android/sample/model/map/NominatimLocationRepositoryTest.kt:138]: This is a well-written test that successfully verifies the failure handling logic of getCurrentLocation. The approach is thorough, and the assertions are well-suited to ensure that the behavior meets expectations. Great job!

**Review by DAFERMOH — COMMENTED** [review]
PR Summary: Location and Permission Handling Tests
Good Coverage: Tests cover key scenarios—permissions granted, denied, and error handling in location retrieval.
Clear Test Structure: Uses Given-When-Then format, making tests readable and easy to follow.
Effective Mocks: Great use of MockK and Mockito.
Suggestions
Helper Methods: Abstract repeated mock setups for readability.
Edge Cases: Consider adding edge cases like null values in location data.

**Review by DAFERMOH — APPROVED** [review]
